pkg/frontend/transport/handler_test.go

```diff
@@ -97,3 +98,62 @@ func TestHandler_ServeHTTP(t *testing.T) {
 		})
 	}
 }
+
+func TestHandler_FailedRoundTrip(t *testing.T) {
+	for _, test := range []struct {
+		name                string
+		expectedMetrics     int
+		path                string
+		expectQueryParamLog bool
+		queryErr            error
+	}{
+		{
+			name:                "Failed round trip with context cancelled",
+			expectedMetrics:     0,
+			path:                "/api/v1/query?query=up&time=2015-07-01T20:10:51.781Z",
+			expectQueryParamLog: true,
+			queryErr:            context.Canceled,
+		},
+		{
+			name:                "Failed round trip with no query params",
+			expectedMetrics:     0,
+			path:                "/api/v1/query",
+			expectQueryParamLog: false,
+			queryErr:            context.Canceled,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			roundTripper := roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+				return nil, test.queryErr
+			})
+
+			reg := prometheus.NewPedanticRegistry()
+			logs := &concurrency.SyncBuffer{}
+			logger := log.NewLogfmtLogger(logs)
+			handler := NewHandler(HandlerConfig{QueryStatsEnabled: true}, roundTripper, logger, reg)
+
+			ctx := user.InjectOrgID(context.Background(), "12345")
+			req := httptest.NewRequest("GET", test.path, nil)
+			req = req.WithContext(ctx)
+			resp := httptest.NewRecorder()
+
+			handler.ServeHTTP(resp, req)
+			require.Equal(t, StatusClientClosedRequest, resp.Code)
+
+			count, err := promtest.GatherAndCount(
+				reg,
+				"cortex_query_seconds_total",
+				"cortex_query_fetched_series_total",
+				"cortex_query_fetched_chunk_bytes_total",
+				"cortex_query_fetched_chunks_total",
+			)
+
+			assert.NoError(t, err)
```

Comment:
`assert.NoError` isn't the best idea, since it will let the test continue in the face of an unexpected error state and that'll be chaotic:
```suggestion
			require.NoError(t, err)
```